optuna/structs.py

```diff
@@ -102,6 +102,38 @@ class FrozenTrial(
 
     internal_fields = ['distributions', 'params_in_internal_repr', 'trial_id']
 
+    def _validate(self):
+        # type: () -> None
+
+        if self.datetime_start is None:
+            raise ValueError('`datetime_start` must be set.')
+
+        if self.state.is_finished():
+            if self.datetime_complete is None:
+                raise ValueError('`datetime_complete` must be set for a finished trial.')
+        else:
+            if self.datetime_complete is not None:
+                raise ValueError('`datetime_complete` must not be set for a finished trial.')
+
+        if self.state == TrialState.COMPLETE and self.value is None:
+            raise ValueError('`value` must be set for a complete trial.')
```

Comment:
Could you add type checking to ensure that the type of `value` is `float`.
I think it can refrain from injecting invalid trials which cause following errors:

```console
>>> study = optuna.create_study()
>>> study.inject_trial(value='a', params={'x': 1}, distributions={'x': optuna.distributions.UniformDistribution(0, 2)})
>>> study.optimize(lambda x: x.suggest_uniform('x', 0, 2), n_trials=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/study.py", line 280, in optimize
    self._optimize_sequential(func, n_trials, timeout, catch)
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/study.py", line 493, in _optimize_sequential
    self._run_trial(func, catch)
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/study.py", line 613, in _run_trial
    self._log_completed_trial(trial_number, result)
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/study.py", line 622, in _log_completed_trial
    trial_number, value, self.best_value, self.best_params))
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/study.py", line 183, in best_value
    best_value = self.best_trial.value
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/study.py", line 198, in best_trial
    return self.storage.get_best_trial(self.study_id)
  File "/Users/yanase/pfn/clone/optuna-pr/optuna/storages/base.py", line 186, in get_best_trial
    return min(all_trials, key=lambda t: t.value)
TypeError: '<' not supported between instances of 'float' and 'str'
```

Reply:
[NOTE] We discussed offline and concluded to tackle this issue in another PR (we want to add more validations not only in this place).
Fortunately, `Study._append_trial` became a private method, so this problem can be avoided by careful developers.